### PR TITLE
network: implement interface bonding

### PR DESCRIPTION
**Pull request by Foxboron**
Signed-off-by: Morten Linderud <morten.linderud@mullvad.net>

### CODE REVIEW

**Review by jensdrenhaus — COMMENTED** [review]
I'd suggest making an own type for `BondingMode` (same scheme as for `IPAddrMode` e.g.) That makes it easy to check for the right values during parsing. Further in _hostcfg.go_ add a `func checkBonding(cfg *HostCfg) error` and add it to the `validationSet`. There you can check for empty `BondName` etc. and remove these checks from _network.go_

**Comment by Foxboron** [comment]
Added `goerr113` to the exclude list to the linter as we are doing some weird error wrapping which the golanglint-ci tools doesn't like. Rest should be fine

File: opts/hostcfg.go [hunk 1]
```diff
@@ -82,43 +85,101 @@ func (i *IPAddrMode) UnmarshalJSON(data []byte) error {
...
+	IPAddrMode        IPAddrMode        `json:"network_mode"`
+	HostIP            *netlink.Addr     `json:"host_ip"`
+	DefaultGateway    *net.IP           `json:"gateway"`
+	DNSServer         *net.IP           `json:"dns"`
+	NetworkInterface  *net.HardwareAddr `json:"network_interface"`
+	ProvisioningURLs  *[]*url.URL       `json:"provisioning_urls"`
+	ID                *string           `json:"identity"`
+	Auth              *string           `json:"authentication"`
+	Timestamp         *time.Time        `json:"timestamp"`
+	NetworkInterfaces *[]*string        `json:"network_interfaces"`
```
Comment (jensdrenhaus) [inline opts/hostcfg.go:148]: Now we have `NetworkInterface *net.HardwareAddr` and `NetworkInterfaces *[]*string`, do we need both? Or can we rename one? Might look confusing from a user's perspective.

File: opts/hostcfg.go [hunk 2]
```diff
@@ -82,43 +85,101 @@ func (i *IPAddrMode) UnmarshalJSON(data []byte) error {
...
+	BondingBalancedRR
+)
+
+// String implements fmt.Stringer.
+func (i BondingMode) String() string {
+	return [...]string{"", "balanced-rr"}[i]
+}
+
+// MarshalJSON implements json.Marshaler.
+func (i BondingMode) MarshalJSON() ([]byte, error) {
```
Comment (jensdrenhaus) [inline opts/hostcfg.go:102]: Replace _i_ with _b_ or _bm_

File: opts/hostcfg.go [hunk 3]
```diff
@@ -82,43 +85,101 @@ func (i *IPAddrMode) UnmarshalJSON(data []byte) error {
...
+	if i != BondingUnset {
+		return json.Marshal(i.String())
+	}
+
+	return []byte(JSONNull), nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (i *BondingMode) UnmarshalJSON(data []byte) error {
```
Comment (jensdrenhaus) [inline opts/hostcfg.go:111]: Replace _i_ with _b_ or _bm_

**Review by jensdrenhaus — COMMENTED** [review]
Can you add test cases regarding valid and invalid bonding setting to `TestHostCfgUnmarshalJSON`

File: opts/hostcfg.go (hunk 1 above)
Comment (Foxboron) [reply opts/hostcfg.go:148]: We want to move the "*net.HardwareAddr" stuff to use interface names. Effectively NetwokInterface is ignored when bonding is used. Changing that thought would require more work so we opted for the easy route

Comment (jensdrenhaus) [reply opts/hostcfg.go:148]: ok, got it.

**Comment by jensdrenhaus** [comment]
@Foxboron, after adding the test, we can merge from my side

**Comment by Foxboron** [comment]
@jensdrenhaus It now has tests covering all the errors we can encounter + a valid configuration. Also found some null pointer issues :)

**Review by jensdrenhaus — APPROVED** [review]

**Pull request merged**